### PR TITLE
Structuring 2.0

### LINKED ISSUE #115: [mat-2110] Unify validate and structuring into from_cycler pipelines
## Problem

There is no canonical list of keys required for each step in the pipeline, particularly structuring and validation. This is particularly difficult to do because there are so many different data formats for cyclers. This is a problem because it is unclear what kind of data allows beep to do what. 

For example, the best defined schema is for Arbin files (`arbin_conversion.yaml`).  While #108 lists the columns required for structuring somewhat, when you go to featurize them, it's hard to know what you need down the line. This inconsistency is reflected in the arguments like `eis` to common classes like PCR, since (seemingly?) only Maccor files can contain this data. It's also pretty hard to figure out what the validation requirements are. RCR and PCR are treated as a one-size fits all solution for all the different cyclers, making downstream debugging more complex.

In short, as a developer, following/debugging the pipeline can be somewhat complex because of the diversity of data + cycler capabilities. The pipeline is also fairly brittle with respect to adding new cyclers and new kinds of data.

## Quick n' dirty solution

There must be sets of *internal* column keys required for
- all tasks, regardless of cycler or available metadata (e.g., interpolation-required columns seem common among all PCRs)
- special tasks which work among more than 1 cycler
- special tasks which only work with one kind of cycler (EIS_spectrum + Maccor seems to be an example of this)

The schema for conversion and validation can be updated to a standard format to clarify this.

## More involved solution

The quick n' dirty soluton is subject to the same issues the current implementation is, just less so. 

One comprehensive fix would be with a fairy major refactor (in addition to the quick n dirty solution). 

A base class `BeepPipeline` can contain methods required for **all** cyclers for structuring, validating, and featurizing. It can also enforce subclasses to implement required methods via abstractmethods. 

Each cycler type would inherit `BeepPipeline`:

```python
class ArbinPipeline(BeepPipeline):
```

These subclasses would implement required `from_file` methods taking care of the specific mappings to go from file columns to internal columns defined in the schema. They can also implement specific cases common to a subset of cyclers:

```python
class IndigoPipeline(BeepPipeline, UsesDateTime)
```

They can also implement structuring or validation procedures unique to a particular cycler:

```python
class MaccorPipeline(BeepPipeline)
    ...
    def structure_waveform(...)
    ...
    def eis_spectrum_from_file(...)
```

Data at each step of the pipeline can be gathered from their attributes:

```python
mp = MaccorPipeline.from_file("some_pipeline.file")            # make a pipeline from a file
mp.structure()                                                 # run structuring
mp.validate()                                                  # validate required data
mp.raw_data                                                    # analogous to RawCyclerRun
mp.structured_data                                             # analogous to ProcessedCyclerRun
mp.structure_data_from_eis_file("eis.file")                    # incorporate EIS data into pipeline, not applicable to other cyclers
mp.featurize(common_features=True, specific_features=True)     # features can include common dq features and others specific to this cycler which may not be available to others
```

New cyclers or data formats can be added according to the guidelines set up in `BeepFeaturizer`, giving a clear path to writing custom processing pipelines. This might also solve some maintainability issues

### PR DESCRIPTION
**Pull request by ardunn**
###### Not all changes included, just the biggest ones.

- Implementing base structuring and validation class `BEEPDatapath`. Merges together `RawCyclerRun` and `ProcessedCyclerRun` with clear indication of when things are being computed (e.g., `obj.structure()` does the heavy interpolation, `obj.validate()` returns a bool for if raw data is valid or not, no running of expensive code in `__init__`s). Closes #115 
- Particular cyclers inherit this base class; this makes adding new cyclers tractable and self-contained, can be done for a new cycler with very few lines of code. Addresses problems in #115 

- Add methods:
    - `structure`: takes over the functionality of to/from_raw_cycler_run methods
    - `autostructure`: runs determine_structuring_parameters, then structures.
- Validation routines are incorporated in a `.validate()` method.
- Add more serialization support with `from_json_file` and `to_json_file` for storing and loading the raw data in/from a file (i.e., directly after ingestion) as well as all metadata and structured data; should be backwards compatible with all previously serialized `ProcessedCyclerRun` json. I.e., if you to_json_file a datapath, **ALL** of the data is stored, including paths of all source files (including eis).
- Add a `paths` attribute to keep track of original data, metadata, and processed data file locations (naively, at least). This means all the file paths related to a datapath are stored within the object. This includes things like EIS, if available
- Tests run a bit (~10%) faster

New usage looks like:
```

arbin = ArbinDatapath.from_file(my_file)

arbin.validate()
arbin.autostructure()

arbin.raw_data
>>> outputs the data equivalent of what is stored in RawCyclerRun currently


is_valid = arbin.validate()

arbin.structured_data
arbin.structured_summary
>>> outputs the data equivalents of what are stored in ProcessedCyclerRun currently

arbin.diagnostic_data
arbin.diagnostic_summary
>>> outputs the diagnostic data currently stored in ProcessedCyclerRun

arbin.to_json(some_file)
>>> writes json to disk in identical format to current beep, to maintain compatibility.

```

Adding new datapaths is now a bit less convoluted, since you can inherit from `BEEPDatapath`. For example, the complete code for implementing `ArbinDatapath` is a single `from_file` method:

```python
class ArbinDatapath(BEEPDatapath):
    @classmethod
    def from_file(cls, path, metadata_path=None):
        data = pd.read_csv(path)
        data.rename(str.lower, axis="columns", inplace=True)
        for column, dtype in ARBIN_CONFIG["data_types"].items():
            if column in data:
                if not data[column].isnull().values.any():
                    data[column] = data[column].astype(dtype)
        data.rename(ARBIN_CONFIG["data_columns"], axis="columns", inplace=True)
        metadata_path = metadata_path if metadata_path else path.replace(".csv", "_Metadata.csv")

        if os.path.exists(metadata_path):
            metadata = pd.read_csv(metadata_path)
            metadata.rename(str.lower, axis="columns", inplace=True)
            metadata.rename(ARBIN_CONFIG["metadata_fields"], axis="columns", inplace=True)
            # Note the to_dict, which scrubs numpy typing
            metadata = {col: item[0] for col, item in metadata.to_dict("list").items()}
        else:
            logger.warning(f"No associated metadata file for Arbin: "
                           f"'{metadata_path}'. No metadata loaded.")
            metadata = {}

        # standardizing time format
        data["date_time_iso"] = data["date_time"].apply(
            lambda x: datetime.utcfromtimestamp(x).replace(
                tzinfo=pytz.UTC).isoformat()
        )
        paths = {
            "raw": path,
            "metadata": metadata_path if metadata else None
        }
        return cls(data, metadata, paths)
```

##### `structure` module --> `structure` subpackage
The package structure (no pun intended) of `structure` now separates the `structure.py` module into several  modules:
```
# Previously, everything in one huge 2200 line file
├── structure.py
```

```
# Now, a subpackage with separation
├── structure
│   ├── __init__.py
│   ├── arbin.py
│   ├── base.py
│   ├── base_eis.py
│   ├── biologic.py
│   ├── cli.py
│   ├── indigo.py
│   ├── maccor.py
│   └── neware.py
```

### Other changes (biggest, not comprehensive)
- Incorporated interpolate_cycles tests into 1 test, should run 3x as fast now
- Added property `paused_intervals` to all datapaths, which uses `get_max_paused_over_threshold` to return the paused intervals instead of the user having to do the `df.apply` themselves. Also, adds test for the negative case where the run was not paused. 
- Added some improved logging capabilities (i.e., report the actual parameters `determine_structuring_parameters` has, well, determined)
- Consolidated some boilerplate, various syntax improvements
- Makes structuring/summarization robust to missing temperature column without necessarily imputing it all as nan
- Add ability to easily remove previous structuring with `.unstructure()` to restructure with new parameters. Does not apply to legacy processed json.
- removed numpy binary save/load as @patrickherring-TRI said it is no longer needed
- `EISpectrum` is now an abstract base class which can implement a `from_file`. Maccor EIS are now a single child class of this ABC implemented inside of `MaccorDatapath`. So now all Maccor-specific ingestion is inside of `MaccorDatapath`, not EISpectrum. Similar to the structure of datapaths, implementing new EIS ingestion required only inheriting from `EISpectrum` and creating a `from_file` method.
- moved eis tests using specific files to the datapath tests corresponding to those files: `EISpectrumTest.test_from_maccor` --> `TestMaccor.test_eis`
- added tests for new eis base classes, including MSONable serialization
- merged capability of PCR`.auto_load` (previously used nowhere else in the code) and `RCR.from_file` for automatic name recognition of files, now is a function in `structure.cli`
- removed boilerplate from `test_json_processing`
- `auto_load_processed` loads any processed beep file, whether serialized with legacy beep or not
- `auto_load loads` any raw cycler file, whether serialized with legacy beep or not.

### Other notes
- I have manually inspected all the images generated by `test_get_diagnostic` (name identical in new structuring, except it is in the `MaccorDatapath` test now) to make sure they are identical to the original beep structuring.
- Since many structuring tests have been moved around or changed, the original test each was based on is commented above the test.
- New attribute names have corresponding comments for their old attribute names (for often used/major attrs):
> PCR.cycles_interpolated = BEEPDatapath.structured_data
> PCR.summary = BEEPDatapath.structured_summary
> PCR.diagnostic_interpolated = BEEPDatapath.diagnostic_data
> PCR.diagnostic_summary = BEEPDatapath.diagnostic_summary (same name)
> RCR.data = BEEPDatapath.raw_data

- The mapping of old structuring methods to new structuring methods is summarized in [this sheet](https://docs.google.com/spreadsheets/d/1wRgN8mXDCF_WiFc5cvgAnD4wvi36zle1j9HM5NXk2oE/edit?usp=sharing). Alternatively, a picture:
- 
image
 

###### Some are due to this PR, most are pre-existing
- ~~`determine_structuring_parameters` seems to require the raw file to work, as well as the BEEP_PROCESSING_DIR env var set; can we make it work with the metadata/data loaded in memory and without this env var? **This is especially important since autostructuring will not work without the env var being specified, as there are no parameters with which to determine structuring parameters!!!**~~ moved to #202
- ~~`autostructure` is used in tests, but needs some dedicated tests~~ moved to #201
- ~~**we should be able to include validation entirely in the structuring datapaths**, not as a separate module.~~ moved to #200
- ~~unrelated to this PR: it seems big file tests are actually never run, since github workflow uses `BEEP_BIG_TESTS` while the actual files use `BIG_FILE_TESTS` env vars~~ fixed by @patrickherring-TRI in #196
- ~~put conversion schemas for each datapath _inside_ the datapath itself rather than get them from constants such as `MACCOR_CONFIG` in `conversion_schemas`. **The idea is that everything specific to a cycler with go in that cycler's structuring file, e.g., all maccor config goes in `maccor.py`***~~ moved to #202
- ~~`step_is_waveform` should be implemented on a datapath-by-datapath basis, not as in `structuring.base`~~ moved to #204
- ~~MaccorDatapath needs specific logic for eis as/from_dict, currently it is not saved*~~ moved to #205
- ~~SimpleValidator is rarely used?~~ moved to #202
- ~~`auto_load` test could be more comprehensive*~~ moved to #203
- ~~Featurization module has massively confusing `from_run`/`from_processed_cycler_run`/`features_from_processed_cycler_run` syntax, should only be one and should be more clear~~ moved to #207
- ~~Featurization `validate_data` methods have EXTREME boilerplate~~ moved to #207
- ~~Tutorial needs an update~~ done in #199
- ~~**Docs need to be much more comprehensive to reflect new structuring syntax.**~~ done in #199
- ~~All validation is just using arbin validation...~~ moved to #202
- ~~Structuring could be faster by using parallelization on df ops~~ already in #178 
- ~~Logging is anemic~~ moved to #209
    - ~~A warning should be logged if determine_structuring_parameters does not find a diagnostic~~ moved to #209
- ~~`TestStructuringUtils` probably shouldn't be in structuring tests~~ moved to #206
- ~~Test files are not organized by module, makes finding where something is used very difficult~~ moved to #206
- ~~Use `json.gz` compression with monty to reduce structured json sizes~~ moved to #208
- ~~`auto_load_processed` needs tests~~ moved to #203
- ~~diagnostic info (i.e., diagnostic_available) should be an instance attribute of `BEEPDatapath`s which is set upon structuring and is default faise or throws error.~~ moved to #210
- ~~`BEEPDatapath` tests need a dedicated diagnostic interpolation and summary test~~ moved to #210

### CODE REVIEW

**Comment by ardunn** [comment]
Note: I'm keeping all the new code in a file structure2.py until I can substitute it into the current tests and everything works; at that time I'll move the new structuring code into the structure.py file

**Comment by ardunn** [comment]
> Work in progress.
> 
> ### Big items:
> * Implementing base structuring and validation class `BEEPDatapath`. Merges together `RawCyclerRun` and `ProcessedCyclerRun` with clear indication of when things are being computed (e.g., `obj.structure()` does the heavy interpolation, no running of expensive code in `__init__`). Closes #115
> * Particular cyclers inherit this base class; this makes adding new cyclers tractable and self-contained, can be done for a new cycler in ~100 lines of code or less. Addresses problems in #115
> * Add method `autostructure` that takes over the functionality of to/from_raw_cycler_run methods
> * ~Merging validation routines into the structuring module~
> * Add more serialization support for storing and loading the raw data in/from a file (i.e., directly after ingestion) as well as all metadata and structured data; should be backwards compatible with all previously serialized `ProcessedCyclerRun` json.
> 
> New usage will look something like:
> 
> ```
> 
> arbin = ArbinDatapath.from_file(my_file)
> 
> arbin.validate()
> arbin.autostructure()
> 
> arbin.raw_data
> >>> outputs the data equivalent of what is stored in RawCyclerRun currently
> 
> arbin.structured_data
> arbin.summary
> >>> outputs the data equivalents of what are stored in ProcessedCyclerRun currently
> 
> arbin.diagnostic_data
> arbin.diagnostic_summary
> >>> outputs the diagnostic data currently stored in ProcessedCyclerRun
> 
> arbin.to_json(some_file)
> >>> writes json to disk in identical format to current beep, to maintain compatibility.
> ```
> 
> Adding new datapaths is now a bit less convoluted, since you can inherit from `BEEPDatapath`. For example, the complete code for implementing `ArbinDatapath` is a single `from_file` method:
> 
> ```python
> class ArbinDatapath(BEEPDatapath):
>     @classmethod
>     def from_file(cls, path, metadata_path=None):
>         data = pd.read_csv(path)
>         data.rename(str.lower, axis="columns", inplace=True)
>         for column, dtype in ARBIN_CONFIG["data_types"].items():
>             if column in data:
>                 if not data[column].isnull().values.any():
>                     data[column] = data[column].astype(dtype)
>         data.rename(ARBIN_CONFIG["data_columns"], axis="columns", inplace=True)
>         metadata_path = metadata_path if metadata_path else path.replace(".csv", "_Metadata.csv")
> 
>         if os.path.exists(metadata_path):
>             metadata = pd.read_csv(metadata_path)
>             metadata.rename(str.lower, axis="columns", inplace=True)
>             metadata.rename(ARBIN_CONFIG["metadata_fields"], axis="columns", inplace=True)
>             # Note the to_dict, which scrubs numpy typing
>             metadata = {col: item[0] for col, item in metadata.to_dict("list").items()}
>         else:
>             logger.warning(f"No associated metadata file for Arbin: "
>                            f"'{metadata_path}'. No metadata loaded.")
>             metadata = {}
> 
>         # standardizing time format
>         data["date_time_iso"] = data["date_time"].apply(
>             lambda x: datetime.utcfromtimestamp(x).replace(
>                 tzinfo=pytz.UTC).isoformat()
>         )
>         paths = {
>             "raw": path,
>             "metadata": metadata_path if metadata else None
>         }
>         return cls(data, metadata, paths)
> ```
> 
> ### Other changes
> * Incorporated interpolate_cycles tests into 1 test, should run 3x as fast now
> * Fixed binary loading/saving test, previously loaded the into the same object rather than from a new object
> * Added property paused_intervals to all datapaths, which uses get_max_paused_over_threshold to return the paused intervals instead of the user having to do the `df.apply` themselves. Also, adds test for the negative case where the run was not paused.
> * Test _should_ run a bit faster as there are fewer repeated calls to methods of RCR/PCR/BEEPDatapaths
> * Added some improved logging capabilities (i.e., report the actual parameters `determine_structuring_parameters` has, well, determined)
> * Makes structuring/summarization robust to missing temperature column
> * Add a `paths` attribute to keep track of original data, metadata, and processed data file locations (naively, at least). This means all the file paths related to a datapath are stored within the object. This includes things like EIS, if available.
> * Add ability to easily remove previous structuring with `.unstructure()` to restructure with new parameters. Does not apply to legacy processed json.
> * removed numpy binary save/load as @patrickherring-TRI said it is no longer needed
> * `EISpectrum` is now an abstract base class which can implement a `from_file`. Maccor EIS are now a single child class of this ABC implemented inside of `MaccorDatapath`. So now all Maccor-specific ingestion is inside of `MaccorDatapath`, not EISpectrum. Similar to the structure of datapaths, implementing new EIS ingestion required only inheriting from `EISpectrum` and creating a `from_file` method.
> 
> ### Areas for improvement/make new GH issues
> * `determine_structuring_parameters` seems to require the raw file to work, as well as the BEEP_PROCESSING_DIR env var set; can we make it work with the metadata/data loaded in memory and without this env var?
> * `temperature` seems to be an implicitly required key of most raw data, especially in `get_summary`. Can we just impute temperature as 0.0s or NaNs?
> * `autostructure` needs some dedicated tests
> * we should be able to include validation in the structuring datapaths
> * unrelated to this PR: it seems big file tests are actually never run, since github workflow uses `BEEP_BIG_TESTS` while the actual files use `BIG_FILE_TESTS` env vars
> * since each datapath is it's own class now, **there is no automatic file recognition**. In the CLI we can migrate that code so that the correct datapath is returned, but as of now it doesn't have a place in the new structuring
> * put configs for each datapath _inside_ the datapath itself rather than get them from constants such as `MACCOR_CONFIG`

**Comment by ardunn** [comment]
I'm pinging at least 2 of @patrickherring-TRI @chirranjeevigopal-TRI @JosephMontoya-TRI to review before merging, since this PR is colossal. 


Future PRs will be somewhat less colossal.

File: beep/structure/cli.py
Comment (patrickherring-TRI) [inline beep/structure/cli.py:131]: @ardunn Should this be `autostructure`?

Comment (ardunn) [reply beep/structure/cli.py:131]: Yes. I'll push a commit to fix that now.

**Review by patrickherring-TRI — APPROVED** [review]
LGTM

**Comment by ardunn** [comment]
Ok! I'd like to get one more review before merging, perhaps from @moorepatrick or @chirranjeevigopal-TRI if they have thoughts.

**Comment by chirranjeevigopal-TRI** [comment]
@ardunn I started to look through it yesterday; I will try to move a few things around this afternoon to finish it.

File: beep/structure/base.py
Comment (JosephMontoya-TRI) [inline beep/structure/base.py:569]: Can we add an error message here to indicate if the axis is wrong?  Might help if e.g. someone mispells the axis parameter.

**Review by JosephMontoya-TRI — APPROVED** [review]
This is a huge improvement, great work.  One very minor comment.

File: beep/structure/base.py
Comment (chirranjeevigopal-TRI) [inline beep/structure/base.py:108]: typo in the docstring `(or not is not easily accessed from)`.

File: beep/structure/base.py
Comment (chirranjeevigopal-TRI) [inline beep/structure/base.py:294]: I like the use of the structure decorators!

File: beep/structure/base.py
Comment (chirranjeevigopal-TRI) [inline beep/structure/base.py:338]: Just to make sure I understand, both raw and interpolated data will be serialized together? Does this slow down the de-serialization? @patrickherring-TRI, would it be helpful to have an option to 'chunk' the deserialization to only load processed or raw data attributes, for instance? Maybe the performance difference is negligible, idk..

Comment (chirranjeevigopal-TRI) [reply beep/structure/base.py:338]: I ask, cause quite often, we don't end up dealing with the raw time series data once it is structured. I imagine we can drop it after loading the object, but I was not sure if it slows things down to re-load it each time.

**Review by chirranjeevigopal-TRI — APPROVED** [review]
Monumental work, thanks @ardunn. I looked through most of the code, and reviewed the tests. A couple of questions and a minor comment, otherwise LGTM!

File: beep/structure/base.py
Comment (ardunn) [reply beep/structure/base.py]: addressed in [0a7fa78](https://github.com/TRI-AMDD/beep/pull/150/commits/0a7fa78f4d39d99469b39718e143d48ad8fe8826)

File: beep/structure/base.py
Comment (ardunn) [reply beep/structure/base.py]: addressed in [9a13ca2](https://github.com/TRI-AMDD/beep/pull/150/commits/9a13ca2f1710a08355a0229a50ae21ec03e1b278)

File: beep/structure/base.py
Comment (ardunn) [reply beep/structure/base.py:293]: Thanks! :)

File: beep/structure/base.py
Comment (ardunn) [reply beep/structure/base.py:337]: Yeah that's a good observation.

The downside of saving `raw_data` is that the saved json are bigger. Depending on the file, 3-4x as large as json from PCR. After compression though, as suggested in #208, `BEEPDatapaths` _with_ raw_data are 2-5x smaller than they were with PCR (not including the raw data). Decompression and loading for the large files I tested (mostly just the test maccor one from AWS) is pretty fast, though not instant. 

The upside is that including the raw data makes the object complete. For example, you can structure a Datapath, save it to disk, load it, _independently restructure any of the most important attributes with different structuring parameters_, save it again, etc. I couldn't see an easy way of doing this when raw data is not saved; maybe the interpolated dfs could be reversed to raw data if needed, but I wasn't smart enough to figure this out. Restructuring files might become important if, for example, something is found wrong with the method for getting diagnostic summaries. Without saving the raw data, you'd have to reload and redo the entire structuring for all files. By including the raw data, you can just change the method, load the previously structured data, regenerate _only_ the diagnostic summaries, and save again to disk.

But to address whether this is actually going to cause a performance problem, I added #211.

**Comment by ardunn** [comment]
Thanks @chirranjeevigopal-TRI @JosephMontoya-TRI ! Once tests pass this final time, I'll merge

**Pull request merged**
